### PR TITLE
nbt v2

### PR DESCRIPTION
**Pull request by ExaInsanity**
Assembly version: 0.1.0

current PR state:
- [x] Abstraction
- [x] NBT token implementation
- [x] NBT reading/writing implementation
- [x] sNBT reading/writing implementation
- [x] docs

### CODE REVIEW

**Comment by ExaInsanity** [comment]
ready for review, @Zefyro, feel free to make your review while i'm writing docs

File: src/.editorconfig [hunk 1]
```diff
@@ -75,6 +75,7 @@ dotnet_naming_style.pascal_case.capitalization = pascal_case
 
 dotnet_naming_style.begins_with_i.required_prefix = I
 dotnet_naming_style.begins_with_i.capitalization = pascal_case
+
```
[the remark this answers is not in the document]
Comment (ExaInsanity) [reply src/.editorconfig:78]: oh right i was working on a few more editorconfig rules

File: src/Helium.Data.Nbt/NbtCompoundToken.cs [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+	/// <summary>
+	/// Represents the direct parent of this token.
+	/// </summary>
+	public IDataToken? ParentToken { get; set; }
+
+	public ICollection<String> Keys => children.Keys;
+
+	public ICollection<IDataToken> Values => children.Values;
+
+	public Int32 Count => this.children.Count;
```
Comment (Zefyro) [inline src/Helium.Data.Nbt/NbtCompoundToken.cs:59]: Am I dumb or why do you need `this.` here?

Comment (Zefyro) [reply src/Helium.Data.Nbt/NbtCompoundToken.cs:59]: or why the other ones don't have `this.` ?

Comment (ExaInsanity) [reply src/Helium.Data.Nbt/NbtCompoundToken.cs:59]: the mistake here is that the others do not have a `this` qualification

Comment (ExaInsanity) [reply src/Helium.Data.Nbt/NbtCompoundToken.cs:59]: resolved

File: src/Helium.Data.Nbt/Serialization/BinaryNbtReader.cs [hunk 3]
```diff
@@ -0,0 +1,970 @@
...
+										 * -
+										 * Re-shuffle the bytes according to the mask defined above; inversing their endianness.
+										 * -
+										 * Store this into our pointer and proceed with the next iteration
+										 * -
+										 * This is available in AVX-512 but .NET does not support AVX-512 yet either; it is not
+										 * widely adopted and therefore a low priority for the .NET team. This code area should be
+										 * updated once those are implemented in System.Runtime.Intrinsics.
+										 */
+										Avx.Store(reshuffled, Avx2.Shuffle(Avx.LoadVector256(
```
Comment (Zefyro) [inline src/Helium.Data.Nbt/Serialization/BinaryNbtReader.cs:250]: so yeah... you explained this to me earlier but I still don't understand this. Guess I'll just assume it works and does what described.

File: src/.editorconfig (hunk 1 above)
[the remark this answers is not in the document]
Comment (ExaInsanity) [reply src/.editorconfig:125]: resolved

File: src/Helium.Data.Nbt/Serialization/BinaryNbtReader.cs (hunk 3 above)
Comment (ExaInsanity) [reply src/Helium.Data.Nbt/Serialization/BinaryNbtReader.cs:250]: 😝

**Comment by ExaInsanity** [comment]
alright, i'm done so far. let me know about any feedback

on an unrelated note, this is pretty massive
image

**Review by Zefyro — APPROVED** [review]
Looking good, even tho I didn't understand some of it.

**Pull request merged**